contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,308 @@
+
+    uint256 internal immutable token0Decimals;
+    uint256 internal immutable token1Decimals;
+
+    uint256 public P; // fixed price param: amount of token1 per amount of token0 in wad
+    uint256 public fee; // fee charged on swap; acts as LP's bid/ask spread
+
+    modifier onlyOwner() {
+        require(msg.sender == owner, "!owner");
+        _;
+    }
+
+    modifier onlyToken(uint8 index) {
+        require(index <= 1, "invalid token index");
+        _;
+    }
+
+    modifier deadlineCheck(uint256 deadline) {
+        require(block.timestamp <= deadline, "block.timestamp > deadline");
+        _;
+    }
+
+    event Quote(uint256 price);
+    event NewSwapFee(uint256 newSwapFee);
+    event TokenSwap(address indexed buyer, uint256 tokensSold, uint256 tokensBought, uint128 soldId, uint128 boughtId);
+    event AddLiquidity(
+        address indexed provider,
+        uint256[] tokenAmounts,
+        uint256[] fees,
+        uint256 invariant,
+        uint256 lpTokenSupply
+    );
+    event RemoveLiquidity(address indexed provider, uint256[] tokenAmounts, uint256 lpTokenSupply);
+
+    constructor(
+        address _owner,
+        address _token0,
+        address _token1
+    ) {
+        factory = msg.sender;
+        owner = _owner;
+        token0 = _token0;
+        token1 = _token1;
+
+        // limit to tokens with decimals <= 18
+        uint256 _token0Decimals = uint256(IERC20Metadata(_token0).decimals());
+        require(_token0Decimals <= 18, "token0 decimals > 18");
+        token0Decimals = _token0Decimals;
+
+        uint256 _token1Decimals = uint256(IERC20Metadata(_token1).decimals());
+        require(_token1Decimals <= 18, "token1 decimals > 18");
+        token1Decimals = _token1Decimals;
+    }
+
+    /// @notice Updates the quote price LP is willing to offer tokens at
+    /// @param _P The new fixed price LP is willing to buy and sell at
+    function quote(uint256 _P) external onlyOwner {
+        require(_P >= PRICE_MIN && _P <= PRICE_MAX, "price out of range");
+        require(_P != P, "same price");
+
+        // set new P price param
+        P = _P;
+
+        emit Quote(_P);
+    }
+
+    /// @notice Updates the fee applied on swaps
+    /// @dev Effectively acts as bid/ask spread for LP
+    /// @param _fee The new swap fee
+    function setSwapFee(uint256 _fee) external onlyOwner {
+        require(_fee <= FEE_MAX, "fee > max");
+        fee = _fee;
+        emit NewSwapFee(_fee);
+    }
+
+    /// @notice Adds liquidity to pool
+    /// @param amounts The token amounts to add in token decimals
+    /// @param minToMint The minimum amount of liquidity to be minted
+    /// @param deadline The deadline before which liquidity must be added
+    function addLiquidity(
+        uint256[] calldata amounts,
+        uint256 minToMint,
+        uint256 deadline
+    ) external onlyOwner deadlineCheck(deadline) returns (uint256 minted_) {
+        require(amounts.length == 2, "invalid amounts");
```

Comment:
I would add the check if the price is non-zero here to ensure that LP doesn't add liquidity without setting the price first.